Implement nblinkgallery directive

adonath opened the pull request:
This PR addresses issue #651 and includes the following changes:
- Implement new `NbLinkGallery` class and `nblinkgallery` directive
- Implement `GalleryLinks` node
- Adapt `doctree_resolve` to re-use the `GalleryNode` for `GalleryLinks` as well
- Add minimal docs example to `a-normal-rst-file.rst`

I consider this still as a draft. I'm happy to receive any comments for improvement, especial reduction of code duplication and improvement of documentation.

mgeier commented:
Thanks, this looks promising!

The HTML result is broken because of an unrelated problem with the bibliography. I'll try to find and report the problem, but in the meantime, you can just remove the `bibliography` directive during development and it should build fine.

The linkcheck error is also unrelated.

mgeier commented on src/nbsphinx.py:
```diff
@@ -1262,6 +1267,29 @@ def run(self):
+        links = []
+
+        for entry in self.content:
+            filename = (path.parent / entry).resolve()
```
To be consistent with [toctree](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-toctree), the entries should be handled as "document names", i.e. they don't require suffixes (but they are supported for backwards compatibility). Also, if an entry starts with a slash, it should be relative to the source directory. I'm not sure if there is already some Sphinx code that can be re-used for that.

I was wondering if `env.found_docs` can be used here, but maybe this is not populated yet at this point?

mgeier replied:
I didn't test this, but maybe you can use the existing `parse_content()` method?

https://github.com/sphinx-doc/sphinx/blob/f45dc5e1398f7850e23003cac200ad3433da2a78/sphinx/directives/other.py#L79

I guess you would have to override the `run()` method, because the options are not exactly equal.

https://github.com/sphinx-doc/sphinx/blob/f45dc5e1398f7850e23003cac200ad3433da2a78/sphinx/directives/other.py#L55

adonath replied:
I was able to re-use the `TocTree.parse_content()` method, however just by passing the `NbLinkGallery` as `self` to  the `.parse_content()` method, which is somewhat unusual and maybe not very future proof. However it works and simplifies the code, while getting the possibility to work with the document names only. I adapted the documentation accordingly.

mgeier replied:
Why can't you derive `NbLinkGallery` from `sphinx.directives.other.TocTree`?

I guess you would also have to override the `option_spec` (and maybe other things).

mgeier replied:
To answer my own question:

> Why can't you derive `NbLinkGallery` from `sphinx.directives.other.TocTree`?

When doing that, links to the gallery notebooks are included in the TOC, which is not what we want.
I didn't find an elegant/clean solution, but I tried replacing the `toctree` node with a dummy node, which took care of the problem.

My experiments are available at 4bcd8dcc95ad80308cf23d625544cd73e47b1484, which is a bit hacky but I think it has quite good code re-use.

It also enables a few missing directive options and disallows the ones that don't work.

What do you think about it?

mgeier replied:
I just created a little variation, which is hopefully somewhat less hacky: 9615c4e76cbe80f049a463e1dbe20c99125325cd

adonath replied:
Thanks @mgeier! I agree your proposed solution is cleaner and even better re-use of code. How do we proceed maybe attach your commits to this PR? Or open another PR against mine?

mgeier replied:
It would be great if you could continue with this PR.

If you like my commit, feel free to take it into your PR, if you like only some parts, just take some parts. And maybe you find some more things to improve.
You could also try to rebase or squash the commits, which would hopefully make the CI pass.

Once we agree on the functionality, I'd like to have a closer look at the documentation.

mgeier replied:
I have continued working on this in #696.

mgeier commented on doc/a-normal-rst-file.rst:
```diff
@@ -296,3 +296,31 @@ The following example gallery was created using:
+-------------------------
+
+In some case it is desired to just create thumbnail links to existing notebooks,
+already inlcuded in a ``toctree``. This can be used e.g. to link to a subset
```
```suggestion
already included in a ``toctree``. This can be used e.g. to link to a subset
```

mgeier commented on doc/a-normal-rst-file.rst:
```diff
@@ -296,3 +296,31 @@ The following example gallery was created using:
     gallery/*-rst
+
+
+Thumbnail Links Galleries
```
```suggestion
Thumbnail Links Galleries (HTML only)
```

mgeier commented on doc/a-normal-rst-file.rst:
```diff
@@ -296,3 +296,31 @@ The following example gallery was created using:
+.. note::
+
+    The notes regarding LaTeX in :ref:`/subdir/gallery.ipynb`
+    and :ref:`/subdir/toctree.ipynb` also apply here!
```
I don't think those notes apply, because the whole thing doesn't work with LaTeX, right?

mgeier commented on src/nbsphinx.py:
```diff
@@ -2088,63 +2162,57 @@ def doctree_resolved(app, doctree, fromdocname):
-                    app.builder.get_target_uri(fromdocname), '')
 
                 # NB: This is how Sphinx implements the "html_sidebars"
                 #     config value in StandaloneHTMLBuilder.add_sidebars()
```
This comment should stay with the following code.

adonath commented:
Thanks @mgeier  for there review! I addressed all of your comments, please let me know if you have further ones. Currently it seems the latex build fails, I'm not sure what to do about this. Can I test the latex build locally?

adonath commented:
@mgeier Do you have time to look at this again? Thanks!

mgeier commented:
> Can I test the latex build locally?

Yes, sure: https://nbsphinx.readthedocs.io/en/0.8.9/contributing.html#building-the-documentation
I then normally change to `build/sphinx/latex/` and run `latexmk -pdf -pv`

> @mgeier Do you have time to look at this again? Thanks!

I currently have little time and next week I'm on holidays. But after that, I'll have a closer look.

mgeier commented:
I've reported the bibliography-related error there: https://github.com/mcmtroffaes/sphinxcontrib-bibtex/issues/309

mgeier commented:
@adonath I took the liberty of using your work from here and building upon it in #696, I hope you don't mind!

adonath commented:
Thanks @mgeier! Yes that’s totally fine. For our project I ended up using a different solution so I didn’t give this PR any priority…
